WIP: JCLOUDS-892 Compute Abstraction Network and Firewall changes

danbroudy opened the pull request:
This is a WIP branch. 

This branch changes the compute abstraction to use the default network when creating nodes in a group. A user can still create nodes in a different network by creating a network and then passing in the URI through `GoogleComputeEngineTemplateOptions`.

Node deletion and clean up makes no attempt to delete networks. This change encourages users to use the default network. If users want to use a different network then they are in charge of creating and destroying the network.

This PR also addresses the issue of creating one firewall rule per inbound port. 

There is still work to be done. 
WDYT?

andreaturli commented:
Thanks @broudy, at first glance it is a great step forward. I want to test it properly before merging. I should be able to do so during the weekend.

andreaturli commented:
@danbroudy the firewall rules are created correctly, but when I destroy the node, the firewalls are not deleted contextually. Can you  doublecheck that?

danbroudy commented:
Turns out there was a problem extracting the group name from firewall rules with names like `<group-name>-port-<hash> `. I removed port from the name and that should fix the problem. I also updated the mock test that was failing for a similar reason. 

I will update this again later this afternoon with a few more fixes.

danbroudy commented:
rebuild

danbroudy commented:
rebuild please

danbroudy commented:
Updated to remove Network from GoogleComputeEngineTemplateOptions [JCLOUDS-896](https://issues.apache.org/jira/browse/JCLOUDS-896).

danbroudy commented:
Now that there is only one firewall per group we could remove the `FirewallTagNamingConvention`.

I would probably implement this by creating one firewall rule with a `GroupNamingConvention.sharedNameForGroup()` for the firewall name and the group name as a target tag. 

This is problematic if users create multiple groups with the same name and different inbound ports.

danbroudy commented:
I would do that update in a subsequent PR. 

This PR is ready for review.

nacx commented on google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineService.java:
```diff
@@ -125,37 +124,30 @@
 
       for (Firewall firewall : concat(firewallApi.list())) {
-         if (firewall == null || !firewall.network().equals(network.selfLink())) {
+         if (firewall == null) {
```
Can this be true at some point? Does the API return `null` objects int he returned Iterable?

danbroudy replied:
It should not and I was unable to reproduce. Removing this check.

nacx commented on google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java:
```diff
@@ -131,12 +130,20 @@
+      else {
+         Iterator<String> iterator = networks.iterator();
+         networkName = nameFromNetworkString(iterator.next());
+         assert !iterator.hasNext() : "Error: Please specify only one network in TemplateOptions when using GCE.";
```
Out of curiosity: why throwing an AssertionError instead of an IllegalArgumentException, or any other runtime exception?

danbroudy replied:
Updating to IllegalArgumentException that is more applicable in this case.

nacx commented on google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java:
```diff
@@ -131,12 +130,20 @@
+         assert !iterator.hasNext() : "Error: Please specify only one network in TemplateOptions when using GCE.";
+      }
+      Network network = api.networks().get(networkName);
+      return checkNotNull(network, "Error: no network with name %s was found", networkName);
```
Would it be more convenient to propagate an IllegalArgumentException instead of a NPE here?

danbroudy replied:
Agreed. Updating.

nacx commented:
Cool @danbroudy! Just a couple minor comments, but lgtm! Thanks!

andreaturli commented on google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java:
```diff
@@ -155,32 +162,70 @@ private void getOrCreateFirewalls(GoogleComputeEngineTemplateOptions templateOpt
+      String name = naming.name(ports);
+      Firewall firewall = firewallApi.get(name);
+      AtomicReference<Operation> operation = null;
+      if (firewall == null){
```
[minor]: add whitespace

What happens if firewall is not null?

danbroudy replied:
If firewall is not null then the specific firewall we are about to create already exists. We are appending a hash to the end of the firewall name to both make the name unique and indicate which ports it opens to tcp and udp. If a firewall with the name already exists then we assume it opens the correct ports and has the correct tag.

andreaturli commented on google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineService.java:
```diff
@@ -125,37 +124,30 @@
       Set<String> orphanedGroups = findOrphanedGroups.apply(deadNodes);
       for (String orphanedGroup : orphanedGroups) {
-         cleanUpNetworksAndFirewallsForGroup(orphanedGroup);
+         cleanUpFirewallsForGroup(orphanedGroup);
```
this code never gets called, I suspect there is a bug in BaseComputeService, I'll let you know soon

danbroudy replied:
It appears to be called when I run the tests.

danbroudy commented:
any update here?

nacx commented:
Thanks for the ping @danbroudy :) lgtm! @andreaturli good to merge?

andreaturli commented:
+1

nacx commented:
About to merge. Just a note: the tests failed initially because I didn't have a "default" network. I just created it manually. It is very possible that I manually removed that network in previous cleanups, so just to confirm: is that "default" network created by default in GCE projects?

danbroudy commented:
Yes, when you create a new project the default network is created for you.

nacx commented:
Thanks for clarifying and for refactoring this!. Pushed to master as [cd16d593](http://git-wip-us.apache.org/repos/asf/jclouds-labs-google/commit/cd16d593).
